contrib/side-cars/fee-estimate.py

```diff
@@ -0,0 +1,188 @@
+"""
```

Comment:
I'm not opposed to just deleting this from the PR, but I don't think it's worth it to add this functionality to the bash script given how clunky it'll be to add that in bash.

This script is something I began putting together after being asked by @wileyj and your bash PR (https://github.com/stacks-network/stacks-core/pull/4077) beat this to the punch. I should have made an issue for it when I got the request. That said, this will work on any platform whereas the bash script won't. I talked with Jesse about whether I should bother submitting this:

> Ashton: Should I bother to submit it?
> Jesse Wiley: totally
> Jesse Wiley: ... I think it's worth having some options. not everyone has the tools to run the bash script (i.e. windows machines)

The exponential backoff is just for APIs throttling the script which I observed in practice, but if any of the APIs fail this script will keep running with the mean of the other values; all of them need to fail for this script to fail. I try to not have a single API endpoint be a final point of fault (though that's the case with the `stacks-block-delay-event-trigger.py` script).

Reply:
Also; this script could run on the order of seconds but I'd suggest the order of around 30 seconds or even a minute given the current block times.x

Reply:
> I don't think it's worth it to add this functionality to the bash script given how clunky it'll be to add that in bash.

I suppose this is a matter of taste, but I don't think bash is clunky at all when it's used in the way it's intended to be used.  Here, the majority of the complexity is in starting subprocesses and piping their outputs into other programs' inputs, which is precisely the task that bash excels at.

Also, consider the deployment story for these two side-cars.  This side-car is meant to run in the same container (or host) as the Stacks node.  We don't expect very much from the host right now; we only need a barebones Linux distro.  While this script requires a couple programs that may not ship with the distro by default (e.g. `curl`, `jq`), they're readily available via the system package manager, they have a small disk and RAM footprint, and getting them installed is almost never going to be a hassle.

By contrast, Python 3.x is a somewhat large thing to install, and you'd need to go and grab `pip` in order to fetch and install `backoff_utils`.  Setting up the host to run the side-car incurs the added complexity of installing a whole language runtime with its language-specific package manager, and then using that language-specific package manager to go and grab more 3rd party libraries.  There's a lot more that can go wrong with this, especially since several popular Linux distros have strong and conflicting opinions about how to install Python packages (Debian is especially guilty of this).

So until we find ourselves in a situation where we cannot use bash plus the slew of commonly-available command-line utilities that are readily available in pretty much every Linux distro, I'm in favor of keeping the bash side-car.

Reply:
> this will work on any platform whereas the bash script won't.

The bash script will work on any platform.  Linux obviously supports this; Mac OS ships with all the right tools as well; Windows has WSL.  We don't support any other platforms.

Reply:
Okay, I think that these are fair points and am not particularly attached to this script. That said, I would like to keep the python approach for the following script because that's not one we expect for most people to use.

Reply:
> It's a lot easier to introduce a security impactful bug in bash than it is in python, at least from what I've seen.

> With all things there's a "if you use it right then it's good," but to your point about !unsafe code in rust "the vast majority of people who think they can write unsafe code safely are not correct".

This conclusion about `unsafe` Rust code, as well as my insistence on avoiding it, is backed by empirical data gathered for over 10 years from both Google and Microsoft -- about 70% of bugs are memory-safety bugs in languages that are not memory-safe.  Do you have empirical data about the frequency, types, and impacts of bugs in bash scripts versus Python?  Not trying to shut you down; I'm genuinely curious, and will gladly retract my bash script if it can be shown that shell scripting is inherently less reliable or secure.

> In this case I see no significant reason to alter the bash script, but I think we should (at some point) discuss alternatives to bash scripts that contributors can alter more easily and with less risk.

If you feel strongly about this, then it's fine by me to include your `fee-estimate.py` script as-is.